rspace/src/main/scala/coop/rchain/rspace/history/RadixTree.scala

```diff
@@ -0,0 +1,850 @@
+    }
+
+    val emptyResult = (emptyExportData, none).pure
+
+    for {
+      rootNodeSerOpt <- getNodeDataFromStore(rootHash)
+      r              <- rootNodeSerOpt.map(doExport).getOrElse(emptyResult)
+    } yield r
+  }
+
+  /**
+    * Radix Tree implementation
+    */
+  class RadixTreeImpl[F[_]: Sync: Parallel](store: KeyValueTypedStore[F, ByteVector, ByteVector]) {
+
+    /**
+      * Load and decode serializing data from KVDB.
+      */
+    private def loadNodeFromStore(nodePtr: ByteVector): F[Option[Node]] =
+      store.get1(nodePtr).map(_.map(Codecs.decode))
+
+    /**
+      * Cache for storing read and decoded nodes.
+      *
+      * Cache stores kv-pairs (hash, node).
+      * Where hash - Blake2b256Hash of serializing nodes data,
+      *       node - deserialized data of this node.
+      */
+    private val cacheR: TrieMap[ByteVector, Node] = TrieMap.empty
+
+    /**
+      * Load one node from [[cacheR]].
+      *
+      * If there is no such record in cache - load and decode from KVDB, then save to cacheR.
+      * If there is no such record in KVDB - execute assert (if set noAssert flag - return emptyNode).
+      */
+    def loadNode(nodePtr: ByteVector, noAssert: Boolean = false): F[Node] = {
+      def errorMsg(): Unit = assert(noAssert, s"Missing node in database. ptr=${nodePtr.toHex}.")
+      val cacheMiss = for {
+        storeNodeOpt <- loadNodeFromStore(nodePtr)
+        _            = storeNodeOpt.map(cacheR.update(nodePtr, _)).getOrElse(errorMsg())
+      } yield storeNodeOpt.getOrElse(emptyNode)
+      for {
+        cacheNodeOpt <- Sync[F].delay(cacheR.get(nodePtr))
+        res          <- cacheNodeOpt.map(_.pure).getOrElse(cacheMiss)
+      } yield res
+    }
+
+    /**
+      * Clear [[cacheR]] (cache for storing read nodes).
+      */
+    def clearReadCache(): Unit = cacheR.clear()
+
+    /**
+      * Cache for storing serializing nodes. For subsequent unloading in KVDB
+      *
+      * Cache stores kv-pairs (hash, bytes).
+      * Where hash -  Blake2b256Hash of bytes,
+      *       bytes - serializing data of nodes.
+      */
+    private val cacheW: TrieMap[ByteVector, ByteVector] = TrieMap.empty
+
+    /**
+      * Serializing and hashing one [[Node]].
+      *
+      * Serializing data load in [[cacheW]].
+      * If detected collision with older cache data - executing assert
+      */
+    def saveNode(node: Node): ByteVector = {
+      val (hash, bytes) = hashNode(node)
+      // Collision alarm
+      def generateErrorMsg(v: Node): Unit = assert(
+        v == node,
+        s"Collision in cache: record with key = ${hash.toHex} has already existed."
+      )
+      cacheR.get(hash).map(generateErrorMsg).getOrElse(cacheR.update(hash, node))
+      cacheW.update(hash, bytes)
+      hash
+    }
+
+    /**
+      * Save all cacheW to KVDB
+      *
+      * If detected collision with older KVDB data - execute assert
+      */
+    def commit(): F[Unit] = {
+      val kvPairs = cacheW.toList
```

Comment:
This expression will be executed when `commit` function will be called but not as part of `F` execution. This means cacheW will be read only once although commit executed multiple times.

To include it as part of _for_ comprehension it must be _suspended_.
`kvPairs <- Sync[F].delay(cacheW.toList)`

---

This also means that function can defined without parenthesis `()` because it's like a value (`F` value) that can be executed multiple times whenever included as part of `F` expression.

Each time _for_ comprehension is executed, internal _for_ comprehension inside _commit_ will be executed.
```scala
for {
  _ <- commit
 ...
} yield ...
```